# Metabolic Syndrome Prediction | 1. Dataset Exploration

-> Load the dataset

-> Explore and confirm features and label(s) of this dataset

-> Explore size/shape of dataset

-> Investigate data type of features and labels and chose any better option for a 
particular column for data type if possible

-> Calculate the memory usage differences

-> Explore the statistical facts like mean, median, x percentiles of the columns

## 1. Load the dataset

In [1]:
import pandas as pd
import warnings

warnings.filterwarnings("ignore")

df = pd.read_csv("Metabolic Syndrome.csv") # reading in the csv file using pandas, similar implementation cn be done for polars (lazy reading)

---

## 2. Explore and confirm features and label(s) of this dataset

In [2]:
df # displaying the dataset

seqn  Age     Sex  Marital  Income         Race  WaistCirc   BMI  \
0     62161   22    Male   Single  8200.0        White       81.0  23.3   
1     62164   44  Female  Married  4500.0        White       80.1  23.2   
2     62169   21    Male   Single   800.0        Asian       69.6  20.1   
3     62172   43  Female   Single  2000.0        Black      120.4  33.3   
4     62177   51    Male  Married     NaN        Asian       81.1  20.1   
...     ...  ...     ...      ...     ...          ...        ...   ...   
2396  71901   48  Female  Married  1000.0        Other        NaN  59.7   
2397  71904   30  Female   Single  2000.0        Asian        NaN  18.0   
2398  71909   28    Male   Single   800.0  MexAmerican      100.8  29.4   
2399  71911   27    Male  Married  8200.0  MexAmerican      106.6  31.3   
2400  71915   60    Male   Single  6200.0        White      106.6  27.5   

      Albuminuria  UrAlbCr  UricAcid  BloodGlucose  HDL  Triglycerides  \
0               0     3.88       4.9            92   41             84   
1               0     8.55       4.5            82   28             56   
2               0     5.07       5.4           107   43             78   
3               0     5.22       5.0           104   73            141   
4               0     8.13       5.0            95   43            126   
...           ...      ...       ...           ...  ...            ...   
2396            0    22.11       5.8           152   57            107   
2397            0     2.90       7.9            91   90             91   
2398            0     2.78       6.2            99   47             84   
2399            0     4.15       6.2           100   41            124   
2400            0    12.82       5.2            91   36            226   

      MetabolicSyndrome  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  
...                 ...  
2396                  0  
2397                  0  
2398                  0  
2399                  1  
2400                  1  

[2401 rows x 15 columns]

In [3]:
df.columns # displaying column names/features

Index(['seqn', 'Age', 'Sex', 'Marital', 'Income', 'Race', 'WaistCirc', 'BMI',
       'Albuminuria', 'UrAlbCr', 'UricAcid', 'BloodGlucose', 'HDL',
       'Triglycerides', 'MetabolicSyndrome'],
      dtype='object')

In [4]:
df.head() # displaying first 5 rows

seqn  Age     Sex  Marital  Income   Race  WaistCirc   BMI  Albuminuria  \
0  62161   22    Male   Single  8200.0  White       81.0  23.3            0   
1  62164   44  Female  Married  4500.0  White       80.1  23.2            0   
2  62169   21    Male   Single   800.0  Asian       69.6  20.1            0   
3  62172   43  Female   Single  2000.0  Black      120.4  33.3            0   
4  62177   51    Male  Married     NaN  Asian       81.1  20.1            0   

   UrAlbCr  UricAcid  BloodGlucose  HDL  Triglycerides  MetabolicSyndrome  
0     3.88       4.9            92   41             84                  0  
1     8.55       4.5            82   28             56                  0  
2     5.07       5.4           107   43             78                  0  
3     5.22       5.0           104   73            141                  0  
4     8.13       5.0            95   43            126                  0

In [5]:
df.tail() # displaying last 5 rows

seqn  Age     Sex  Marital  Income         Race  WaistCirc   BMI  \
2396  71901   48  Female  Married  1000.0        Other        NaN  59.7   
2397  71904   30  Female   Single  2000.0        Asian        NaN  18.0   
2398  71909   28    Male   Single   800.0  MexAmerican      100.8  29.4   
2399  71911   27    Male  Married  8200.0  MexAmerican      106.6  31.3   
2400  71915   60    Male   Single  6200.0        White      106.6  27.5   

      Albuminuria  UrAlbCr  UricAcid  BloodGlucose  HDL  Triglycerides  \
2396            0    22.11       5.8           152   57            107   
2397            0     2.90       7.9            91   90             91   
2398            0     2.78       6.2            99   47             84   
2399            0     4.15       6.2           100   41            124   
2400            0    12.82       5.2            91   36            226   

      MetabolicSyndrome  
2396                  0  
2397                  0  
2398                  0  
2399                  1  
2400                  1

---

## 3. Explore size/shape of dataset

In [6]:
df.shape # displaying shape of the dataset

(2401, 15)

---

## 4. Investigate data type of features and labels and chose any better option for a particular column for data type if possible

In [7]:
df.info() # displays information about the dataset, with respective datatypes for each column

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2401 entries, 0 to 2400
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   seqn               2401 non-null   int64  
 1   Age                2401 non-null   int64  
 2   Sex                2401 non-null   object 
 3   Marital            2193 non-null   object 
 4   Income             2284 non-null   float64
 5   Race               2401 non-null   object 
 6   WaistCirc          2316 non-null   float64
 7   BMI                2375 non-null   float64
 8   Albuminuria        2401 non-null   int64  
 9   UrAlbCr            2401 non-null   float64
 10  UricAcid           2401 non-null   float64
 11  BloodGlucose       2401 non-null   int64  
 12  HDL                2401 non-null   int64  
 13  Triglycerides      2401 non-null   int64  
 14  MetabolicSyndrome  2401 non-null   int64  
dtypes: float64(5), int64(7), object(3)
memory usage: 281.5+ KB


From above list we see datatypes include float, int and object. ML models require data to be numerical, so we would have to convert datatype object into float/int to proceed further in the ML pipeline.

Identified columns that need above change: with corresponding encoding techniques

-> Sex : binary encoding : one-hot or label encoder

-> Marital : multi-value encoding : label or ordinal encoder

-> Race : multi-value encoding : label encoder

In [8]:
set(df['Sex']) # unique values to be encoded in Sex

{'Female', 'Male'}

In [9]:
set(df['Marital']) # unique values to be encoded in Marital status

{'Divorced', 'Married', 'Separated', 'Single', 'Widowed', nan}

In [10]:
set(df['Race']) # unique values to be encoded in Race

{'Asian', 'Black', 'Hispanic', 'MexAmerican', 'Other', 'White'}

---

# 5. Calculate the memory usage differences

### pandas

In [11]:
print("Pandas Memory Usage:", df.memory_usage().sum(), 'Bytes')

Pandas Memory Usage: 288248 Bytes


In [12]:
print("Pandas Deep Memory Usage:", df.memory_usage(deep=True).sum(), 'Bytes')

Pandas Deep Memory Usage: 677600 Bytes


### polars

In [13]:
import polars as pl

df1 = pl.read_csv('Metabolic Syndrome.csv')

print("Polars Memory Usage:", df1.estimated_size('kb') , 'KB')

del df1 # deleting polars implementation since pandas implementation is not memory-intensive

Polars Memory Usage: 266.537109375 KB


## memory usage for dataset

### pandas

memory usage: 661.7 KB

lazy read memory usage: 282 KB

### polars

lazy read memory usage: 267 KB

---

## 6. Explore the statistical facts like mean, median, x percentiles of the columns

In [14]:
df.describe(percentiles=[0.1,0.25,0.5,0.75,0.99]) # describes numerical interpretations of the dataset in terms of mean, max, min, quartiles, etc.

# percentiles considered: 10, 25, 50, 75, 99

seqn          Age       Income    WaistCirc          BMI  \
count   2401.000000  2401.000000  2284.000000  2316.000000  2375.000000   
mean   67030.674302    48.691795  4005.253940    98.307254    28.702189   
std     2823.565114    17.632852  2954.032186    16.252634     6.662242   
min    62161.000000    20.000000   300.000000    56.200000    13.400000   
10%    63133.000000    25.000000  1000.000000    78.600000    21.400000   
25%    64591.000000    34.000000  1600.000000    86.675000    24.000000   
50%    67059.000000    48.000000  2500.000000    97.000000    27.700000   
75%    69495.000000    63.000000  6200.000000   107.625000    32.100000   
99%    71830.000000    80.000000  9000.000000   145.165000    49.200000   
max    71915.000000    80.000000  9000.000000   176.000000    68.700000   

       Albuminuria      UrAlbCr     UricAcid  BloodGlucose          HDL  \
count  2401.000000  2401.000000  2401.000000   2401.000000  2401.000000   
mean      0.154102    43.626131     5.489046    108.247813    53.369429   
std       0.422780   258.272829     1.439358     34.820657    15.185537   
min       0.000000     1.400000     1.800000     39.000000    14.000000   
10%       0.000000     3.220000     3.700000     86.000000    36.000000   
25%       0.000000     4.450000     4.500000     92.000000    43.000000   
50%       0.000000     7.070000     5.400000     99.000000    51.000000   
75%       0.000000    13.690000     6.400000    110.000000    62.000000   
99%       2.000000   806.580000     9.400000    278.000000    97.000000   
max       2.000000  5928.000000    11.300000    382.000000   156.000000   

       Triglycerides  MetabolicSyndrome  
count    2401.000000        2401.000000  
mean      128.125364           0.342357  
std        95.322477           0.474597  
min        26.000000           0.000000  
10%        56.000000           0.000000  
25%        75.000000           0.000000  
50%       103.000000           0.000000  
75%       150.000000           1.000000  
99%       511.000000           1.000000  
max      1562.000000           1.000000

only for numerical data*

---

## 7. Finding NULL/NaN count for imputation

In [15]:
df.isnull().sum() # counting null values and summing them for each column

seqn                   0
Age                    0
Sex                    0
Marital              208
Income               117
Race                   0
WaistCirc             85
BMI                   26
Albuminuria            0
UrAlbCr                0
UricAcid               0
BloodGlucose           0
HDL                    0
Triglycerides          0
MetabolicSyndrome      0
dtype: int64

in addition to encoding we would also need imputation on the following columns since these hold null/NaN values which cannot be sent into a ML model:

-> Marital

-> Income

-> WaistCirc

-> BMI

of the above null values we would have to check if the columns are disjoint from others or dependent on another column(s) :

for example, income could be evaluated to depend on Sex and Marital status, or BMI being dependent on Sex, and hence keeping in mind these dependencies we would have to encode values.